### PR TITLE
Vert.x web - support RX routing context

### LINKED ISSUE #1589: vert.x web extension does not support RX routing context
The vert.x web extension does not currently support the reactive version of the RoutingContext. Code such as ;
```
...
import io.vertx.reactivex.ext.web.RoutingContext;
...

public class SimpleResource {

    @Route(path = "/simple", methods = GET)
    void getSimple(RoutingContext context) {
...
    }
}
```

fails with;

```
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:build (default) on project vertx-web-quarkus: Failed to build a runnable JAR: Failed to build a runner jar: Failed to augment application classes: Build failure: Build failed due to errors
[ERROR] 	[error]: Build step io.quarkus.arc.deployment.ArcAnnotationProcessor#build threw an exception: java.lang.IllegalStateException: Route handler business method must accept exactly one parameter of type RoutingContext/RoutingExchange: [io.vertx.reactivex.ext.web.RoutingContext] [method: void getSimple(io.vertx.reactivex.ext.web.RoutingContext context), bean:CLASS bean [types=[io.quarkus.test.SimpleResource, java.lang.Object], qualifiers=[@Default, @Any], target=io.quarkus.test.SimpleResource]
```

We should either;
1) support both `io.vertx.reactivex.ext.web.RoutingContext` and `io.vertx.ext.web.RoutingContext`

or

2) Raise a warning if `io.vertx.reactivex.ext.web.RoutingContext`  is passed as a `@Route` method parameter type

### PR DESCRIPTION
**Pull request by mkouba**
- resolves #1589
- also validate method return type

### CODE REVIEW

File: extensions/vertx-web/deployment/src/test/java/io/quarkus/vertx/web/SimpleRouteTest.java [hunk 1]
```diff
@@ -52,6 +53,12 @@ void hello(RoutingContext context) {
             context.response().setStatusCode(200).end("Hello " + (name != null ? name : "world") + "!");
         }
 
+        @Route(path = "/rx-hello")
+        void rxHello(io.vertx.reactivex.ext.web.RoutingContext context) {
+            String name = context.request().getParam("name");
+            context.response().setStatusCode(200).end("Hello " + (name != null ? name : "world") + "!");
+        }
+
```
Comment (johnaohara) [inline extensions/vertx-web/deployment/src/test/java/io/quarkus/vertx/web/SimpleRouteTest.java]: We probably want a RoutingExchangeImpl test case as well

**Review by johnaohara — CHANGES_REQUESTED** [review]
@mkouba I have commented on some of the changes, thanks

File: extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java [hunk 2]
```diff
@@ -169,10 +171,16 @@ public void transform(TransformationContext context) {
         });
     }
 
-    private void validateMethodParameters(BeanInfo bean, MethodInfo method) {
+    private void validateMethod(BeanInfo bean, MethodInfo method) {
+        if (!method.returnType().kind().equals(Type.Kind.VOID)) {
+            throw new IllegalStateException(
+                    String.format("Route handler business method must return void [method: %s, bean:%s",
+                            method.returnType(), method, bean));
+        }
```
[the remark this answers is not in the document]
Comment (mkouba) [reply extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java]: Good catch!

File: extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java [hunk 3]
```diff
@@ -169,10 +171,16 @@ public void transform(TransformationContext context) {
...
         List<Type> params = method.parameters();
         if (params.size() != 1
-                || !(params.get(0).name().equals(ROUTING_CONTEXT) || params.get(0).name().equals(ROUTING_EXCHANGE))) {
+                || !(params.get(0).name().equals(ROUTING_CONTEXT) || params.get(0).name().equals(RX_ROUTING_CONTEXT)
+                        || params.get(0).name().equals(ROUTING_EXCHANGE))) {
             throw new IllegalStateException(String.format(
```
[the remark this answers is not in the document]
Comment (mkouba) [reply extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java]: Not really. We negate the whole expression in the round brackets. But I agree that it does not read well. I'll try to make it more clear.

File: extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java [hunk 4]
```diff
@@ -169,10 +171,16 @@ public void transform(TransformationContext context) {
...
                     "Route handler business method must accept exactly one parameter of type RoutingContext/RoutingExchange: %s [method: %s, bean:%s",
                     params, method, bean));
```
[the remark this answers is not in the document]
Comment (mkouba) [reply extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java]: Yes, another good catch! ;-)

File: extensions/vertx-web/deployment/src/test/java/io/quarkus/vertx/web/SimpleRouteTest.java (hunk 1 above)
Comment (mkouba) [reply extensions/vertx-web/deployment/src/test/java/io/quarkus/vertx/web/SimpleRouteTest.java]: `RoutingExchangeImpl` is already covered by methods accepting `RoutingExchange`.

File: extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (johnaohara) [reply extensions/vertx-web/deployment/src/main/java/io/quarkus/vertx/web/deployment/VertxWebProcessor.java]: Ah, yes I see that now, I missed the opening `(`

File: extensions/vertx-web/deployment/src/test/java/io/quarkus/vertx/web/SimpleRouteTest.java (hunk 1 above)
Comment (johnaohara) [reply extensions/vertx-web/deployment/src/test/java/io/quarkus/vertx/web/SimpleRouteTest.java]: I was thinking that we were missing the negation I mentioned before, and an explicit test would find that missing negation. But agree, another test case would increase test time for not much more coverage

**Comment by mkouba** [comment]
@johnaohara comments addressed ;-)

**Comment by johnaohara** [comment]
@mkouba great, thanks

**Review by johnaohara — APPROVED** [review]

**Review by cescoffier — APPROVED** [review]

**Comment by cescoffier** [comment]
Waiting for CI

**Comment by johnaohara** [comment]
@cescoffier do we need to re-run the quarkusio.quarkus build? it has failed, but there are no doc changes

**Comment by mkouba** [comment]
Looks like some CI failure because the build actually passed, well except the windows machine that was "abandoned" (I have no idea what it means).

**Comment by mkouba** [comment]
Tried to re-run the windows build.

**Comment by johnaohara** [comment]
The first Windows build was "abandoned", but the second build passed. The first Windows build appears to be stuck in an "In progress" state

**Pull request merged**